### PR TITLE
fixed concurrency issue within changeTingType and improved logging

### PR DESCRIPTION
**Pull request by sjka**
relates to #1605

Signed-off-by: Simon Kaufmann <simon.kfm@googlemail.com>

### CODE REVIEW

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java [hunk 1]
```diff
@@ -98,7 +102,11 @@ public void unregisterHandler(Thing thing) {
         ServiceRegistration<ThingHandler> thingHandlerServiceRegistration = thingHandlers
                 .remove(thing.getUID().toString());
         if (thingHandlerServiceRegistration != null) {
+            logger.debug("Removing service registration for handler of thing " + thing.getUID().getAsString());
             unregisterHandler(thingHandlerServiceRegistration);
+        } else {
+            logger.error(MessageFormat.format("There is no service registration for a thing handler of thing {0}.",
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java]: Why do we need to use MessageFormat here?

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingManager.java [hunk 2]
```diff
@@ -278,6 +282,25 @@ public void run() {
...
                 }
 
+                private void waitUntilHandlerUnregistered(final Thing thing, int timeout) {
+                    for (int i = 0; i < timeout / 100; i++) {
+                        if (thing.getHandler() == null && thingHandlers.get(thing.getUID()) == null) {
+                            return;
+                        }
+                        try {
+                            Thread.sleep(100);
+                            logger.debug(MessageFormat.format(
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingManager.java]: Same here, why do we need MessageFormat here?

**Comment by sjka** [comment]
You are right, there was no specific reason to use MessageFormat here. I have updated the PR.

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java [hunk 3]
```diff
@@ -98,7 +101,11 @@ public void unregisterHandler(Thing thing) {
         ServiceRegistration<ThingHandler> thingHandlerServiceRegistration = thingHandlers
                 .remove(thing.getUID().toString());
         if (thingHandlerServiceRegistration != null) {
+            logger.debug("Removing service registration for handler of thing " + thing.getUID().getAsString());
             unregisterHandler(thingHandlerServiceRegistration);
+        } else {
+            logger.error(
+                    "There is no service registration for a thing handler of thing " + thing.getUID().getAsString());
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java]: Don't build / concat strings in the logger arguments, this should be done by the logger itself.
Use '{}' and the arg list.

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java [hunk 4]
```diff
@@ -98,7 +101,11 @@ public void unregisterHandler(Thing thing) {
         ServiceRegistration<ThingHandler> thingHandlerServiceRegistration = thingHandlers
                 .remove(thing.getUID().toString());
         if (thingHandlerServiceRegistration != null) {
+            logger.debug("Removing service registration for handler of thing " + thing.getUID().getAsString());
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java]: Don't build / concat strings in the logger arguments, this should be done by the logger itself.
Use '{}' and the arg list.

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingManager.java [hunk 5]
```diff
@@ -278,6 +282,24 @@ public void run() {
...
 
+                private void waitUntilHandlerUnregistered(final Thing thing, int timeout) {
+                    for (int i = 0; i < timeout / 100; i++) {
+                        if (thing.getHandler() == null && thingHandlers.get(thing.getUID()) == null) {
+                            return;
+                        }
+                        try {
+                            Thread.sleep(100);
+                            logger.debug("Waiting for handler deregistration to complete for thing {}. Took already {}ms.",
+                                    thing.getUID().getAsString(), i * 100);
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingManager.java]: That the way you should use the logger above :wink:

**Comment by sjka** [comment]
err, well... PR updated.

In the meantime, @afuechsel tested this actually fixed his issue:  https://github.com/eclipse/smarthome/issues/1605#issuecomment-224596959

**Pull request merged**